### PR TITLE
Untag dispatch regex

### PR DESCRIPTION
**Pull request by AlexGuteniev**
Towards #189
Separated from #2219

### CODE REVIEW

File: stl/inc/regex [hunk 1]
```diff
@@ -1998,23 +1989,24 @@ private:
...
 #endif // _ENHANCED_REGEX_VISUALIZER
 
-        _Parser<_FwdIt, _Elem, _RxTraits> _Prs(_Traits, _First, _Last, _Flags);
-        _Root_node* _Rx = _Prs._Compile();
-        _Reset(_Rx);
+            _Parser<_InIt, _Elem, _RxTraits> _Prs(_Traits, _First, _Last, _Flags);
+            _Root_node* _Rx = _Prs._Compile();
+            _Reset(_Rx);
+        } else {
+            static_assert(_Is_input_iter_v<_InIt>, "SHould be at least input iterator");
```
Comment (cristeigabriel) [inline stl/inc/regex:2003]: It appears that there's a mistake in casing

File: stl/inc/regex [hunk 2]
```diff
@@ -547,8 +509,38 @@ int _Iter_compare(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _FwdIt2 _Las
...
-    return _Iter_compare2<_Traits>(
-        _Get_unwrapped(_First1), _Get_unwrapped(_Last1), _Get_unwrapped(_First2), _Get_unwrapped(_Last2));
+
+    auto _UFirst1 = _Get_unwrapped(_First1);
+    auto _ULast1  = _Get_unwrapped(_Last1);
+    auto _UFirst2 = _Get_unwrapped(_First2);
+    auto _ULast2  = _Get_unwrapped(_Last2);
+
+    if constexpr (is_pointer_v<decltype(_UFirst1)> && is_pointer_v<decltype(_UFirst2)>) {
+        // _Iter_compare for memory buffer ranges
```
Comment (CaseyCarter) [inline stl/inc/regex:519]: :fire: extraneous comment :fire: (and on 523).

File: stl/inc/regex [hunk 3]
```diff
@@ -1998,23 +1989,24 @@ private:
...
-
-        _Reset(_Str.data(), _Str.data() + static_cast<ptrdiff_t>(_Str.size()), _Flags, forward_iterator_tag{});
-    }
-
-    template <class _FwdIt>
-    void _Reset(_FwdIt _First, _FwdIt _Last, flag_type _Flags, forward_iterator_tag) {
-        // build regular expression from forward iterators
+    void _Reset(_InIt _First, _InIt _Last, flag_type _Flags) {
+        if constexpr (_Is_fwd_iter_v<_InIt>) {
+            // build regular expression from forward iterators
```
Comment (CaseyCarter) [inline stl/inc/regex:1994]: Ditto "merge with other comment and promote to function description".

File: stl/inc/regex [hunk 4]
```diff
@@ -1998,23 +1989,24 @@ private:
...
 #endif // _ENHANCED_REGEX_VISUALIZER
 
-        _Parser<_FwdIt, _Elem, _RxTraits> _Prs(_Traits, _First, _Last, _Flags);
-        _Root_node* _Rx = _Prs._Compile();
-        _Reset(_Rx);
+            _Parser<_InIt, _Elem, _RxTraits> _Prs(_Traits, _First, _Last, _Flags);
+            _Root_node* _Rx = _Prs._Compile();
+            _Reset(_Rx);
+        } else {
+            static_assert(_Is_input_iter_v<_InIt>, "Should be at least input iterator");
```
Comment (CaseyCarter) [inline stl/inc/regex:2003]: "Iterators must be input iterators" would be more consistent with the other PRs.

**Review by CaseyCarter — APPROVED** [review]
Nitpicky cosmetic things; I'll push changes.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for `<regex>` replacing this tag dispatch with `if constexpr`! :joy_cat: :heart_eyes_cat: :tada:

**Pull request merged**